cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.batch.dataset.input;
+
+import co.cask.cdap.api.data.batch.PartitionedFileSetInputContext;
+import co.cask.cdap.api.dataset.lib.PartitionKey;
+import co.cask.cdap.api.dataset.lib.partitioned.PartitionKeyCodec;
+import co.cask.cdap.data2.dataset2.lib.partitioned.PartitionedFileSetDataset;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.lib.input.FileSplit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.net.URI;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A basic implementation of {@link PartitionedFileSetInputContext}.
+ */
+class BasicPartitionedFileSetInputContext extends BasicInputContext implements PartitionedFileSetInputContext {
+
+  private static final Gson GSON =
+    new GsonBuilder().registerTypeAdapter(PartitionKey.class, new PartitionKeyCodec()).create();
+
+  private static final Type STRING_PARTITION_KEY_MAP_TYPE = new TypeToken<Map<String, PartitionKey>>() { }.getType();
+
+  private final Path inputPath;
+  private final String mappingString;
+
+  private PartitionKey partitionKey;
+
+  BasicPartitionedFileSetInputContext(MultiInputTaggedSplit multiInputTaggedSplit) {
+    super(multiInputTaggedSplit.getName());
+
+    InputSplit inputSplit = multiInputTaggedSplit.getInputSplit();
+    if (!(inputSplit instanceof FileSplit)) {
+      // TODO: need to support FileSplit (old version)?
+      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+    }
+    this.inputPath = ((FileSplit) inputSplit).getPath();
+    this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+  }
+
+  @Override
+  public PartitionKey getInputPartitionKey() {
+    if (partitionKey == null) {
+      partitionKey = getPartitionKey(inputPath.toUri());
+    }
+    return partitionKey;
+  }
+
+  private PartitionKey getPartitionKey(URI inputPathURI) {
+    Map<String, PartitionKey> pathToPartitionMapping =
```

Comment:
To tag the input splits with the partition information is tricky because the InputFormat class which returns the actual input splits may return multiple splits per partition key.
So, once we do (pseudocode):
`List<Split> splits = InputFormat#getSplit(partitionKeys)`, its difficult to know which of the splits are for which partitions.